src/engine/sidechain/shoutconnection.cpp

```diff
@@ -0,0 +1,979 @@
+    m_readSema.release();
+}
+
+void ShoutConnection::setOutputFifo(FIFO<CSAMPLE>* pOutputFifo) {
+    Q_UNUSED(pOutputFifo);
+}
+
+FIFO<CSAMPLE>* ShoutConnection::getOutputFifo() {
+    return m_pOutputFifo;
+}
+
+bool ShoutConnection::threadWaiting() {
+    return m_threadWaiting;
+}
+
+void ShoutConnection::run() {
+    QThread::currentThread()->setObjectName(
+            QString("ShoutOutput '%1'").arg(m_pProfile->getProfileName()));
+    qDebug() << "ShoutOutput::run: Starting thread";
+
+#ifndef __WINDOWS__
+    ignoreSigpipe();
+#endif
+
+    VERIFY_OR_DEBUG_ASSERT(m_pOutputFifo) {
+        qDebug() << "ShoutOutput::run: Broadcast FIFO handle is not available. Aborting";
+        return;
+    }
+
+    if (!processConnect()) {
+        m_pProfile->setEnabled(false);
+        errorDialog(tr("Can't connect to streaming server"),
+                m_lastErrorStr + "\n" +
+                tr("Please check your connection to the Internet and verify that your username and password are correct."));
+        return;
+    }
+
+    setStatus(BroadcastProfile::STATUS_CONNECTED);
+
+    while(true) {
+        // Stop the thread if broadcasting is turned off
+        if (!m_pProfile->getEnabled() || !m_pBroadcastEnabled->toBool()) {
+            m_threadWaiting = false;
+            qDebug() << "ShoutOutput::run: Connection disabled. Disconnecting";
+            if(processDisconnect()) {
+                setStatus(BroadcastProfile::STATUS_UNCONNECTED);
+            }
+            setFunctionCode(2);
+            break;
+        }
+
+        setFunctionCode(1);
+        incRunCount();
+        if(!m_readSema.tryAcquire(1, 1000)) {
+            continue;
+        }
+
+        int readAvailable = m_pOutputFifo->readAvailable();
+        if (readAvailable) {
+            setFunctionCode(3);
+            CSAMPLE* dataPtr1;
+            ring_buffer_size_t size1;
+            CSAMPLE* dataPtr2;
+            ring_buffer_size_t size2;
+
+            // We use size1 and size2, so we can ignore the return value
+            (void)m_pOutputFifo->aquireReadRegions(readAvailable, &dataPtr1, &size1,
+                    &dataPtr2, &size2);
+
+            // Push frames to the encoder.
+            process(dataPtr1, size1);
+            if (size2 > 0) {
+                process(dataPtr2, size2);
+            }
+
+            m_pOutputFifo->releaseReadRegions(readAvailable);
+        }
+    }
+
+    qDebug() << "ShoutOutput::run: Thread stopped";
+}
+
+#ifndef __WINDOWS__
+void ShoutConnection::ignoreSigpipe() {
+    // If the remote connection is closed, shout_send_raw() can cause a
+    // SIGPIPE. If it is unhandled then Mixxx will quit immediately.
+#ifdef Q_OS_MAC
+    // The per-thread approach using pthread_sigmask below does not seem to work
+    // on macOS.
+    struct sigaction sa;
+    memset(&sa, 0, sizeof(sa));
+    sa.sa_handler = SIG_IGN;
+    if (sigaction(SIGPIPE, &sa, NULL) != 0) {
+        qDebug() << "EngineBroadcast::ignoreSigpipe() failed";
```

Comment:
Yes, I've used it in BroadcastSettings. What is the difference between qDebug and Logger?